src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,345 @@
+			}
+
+			// must be at least one contributor
+			this.contributors = Common.dedupeSimpleList(contributors);
+			if (this.contributors == null || this.contributors.isEmpty()) {
+				this.errorList.add("at least one contributor must be provided");
+			}
+
+			// must be at least one title
+			this.titles = Common.dedupeSimpleList(titles);
+			if (this.titles == null || this.titles.isEmpty()) {
+				this.errorList.add("at least one title must be provided");
+			}
+		}
+
+		/**
+		 * Sets comments for the resource
+		 * @param comments comments as a string
+		 * @return this builder
+		 */
+		public Builder withComments(final String comments) {
+			this.comments = Common.processMultilineString(comments);
+			return this;
+		}
+
+		/**
+		 * Sets the license for the resource
+		 * @param license license as a string
+		 * @return this builder
+		 */
+		public Builder withLicense(final String license) {
+			this.license = Common.processString(license);
+			return this;
+		}
+
+		/**
+		 * Sets the version
+		 * @param withVersion version string
+		 * @return this builder
+		 */
+		public Builder withVersion(final String version) {
+			this.version = Common.processString(version);
+			return this;
+		}
+
+		/**
+		 * Sets the dates of various resource life cycle events
+		 * @param withDates list of {@link EventDate}s.
+		 * @return this builder
+		 */
+		public Builder withDates(final List<EventDate> dates) {
+			this.dates = Common.dedupeSimpleList(dates);
+			return this;
+		}
+
+		/**
+		 * Sets the funding source(s)
+		 * @param withFunding list of {@link FundingReference}s.
+		 * @return this builder
+		 */
+		public Builder withFunding(final List<FundingReference> funding) {
+			this.funding = Common.dedupeSimpleList(funding);
+			return this;
+		}
+
+		/**
+		 * Sets the related identifiers
+		 * @param withRelatedIdentifiers list of {@link PermanentID}s.
+		 * @return this builder
+		 */
+		public Builder withRelatedIdentifiers(final List<PermanentID> relatedIdentifiers) {
+			this.relatedIdentifiers = Common.dedupeSimpleList(relatedIdentifiers);
+			return this;
+		}
+
+		/**
+		 * Build the {@link CreditMetadata}.
+		 *
+		 * @return the credit metadata.
+		 */
+		public CreditMetadata build() {
+
+			if (version == null && (dates == null || dates.isEmpty())) {
+				errorList.add("must provide either 'version' or one or more 'dates', ideally indicating when the resource was published or when it was last updated");
+			}
+
+			// If the license looks like a URL, ensure it's properly formatted and seems valid.
+			// For these purposes, if it starts with http or contains '://', assume it's an URL.
+			if (license != null && (license.toLowerCase().startsWith("http") || license.contains("://"))) {
+				try {
+					final URL licenseURL = Common.processURL(license, "license");
+					// license field is a string, so convert the URL back to string form
+					license = licenseURL.toURI().normalize().toString();
```

Comment:
toURI can throw an exception that probably isn't going to be very useful for users debugging-wsie

Reply:
Just as a thought, you may want to write some integration tests at the argutils level when coverting the generated classes to these classes to see how hard it is to debug an error deep in the class hierarcy

Reply:
`processUrl` wraps the error in a comforting message saying "Illegal license URL <url>: <specific error>", so they should at least realise that the URL is malformed.

Reply:
Oh, toURI() is already called there, ok.

:+1: